[IMP] queue_job_cron_jobrunner: channel

Koudja opened the pull request:
Implement channel on queue_job_cron_jobrunner

Koudja committed:
[IMP] queue_job_cron_jobrunner: channel

alexandregaldeano commented on queue_job_cron_jobrunner/models/queue_job.py:
```diff
@@ -32,42 +33,77 @@ def _acquire_one_job(self):
+        for queue_job in self.search([("state", "=", "started")]):
+            if not queue_job.channel:
+                continue
+            channels[queue_job.channel] = channels.get(queue_job.channel, 0) + 1
```
**suggestion:** what do you think?
```suggestion
        channels = defaultdict(int)
        for queue_job in self.search([("state", "=", "started")]):
            if not queue_job.channel:
                continue
            channels[queue_job.channel] += 1
```

alexandregaldeano commented on queue_job_cron_jobrunner/models/queue_job.py:
```diff
@@ -32,42 +33,77 @@ def _acquire_one_job(self):
+            result = queue_job
+            break
+        self.flush()
+        self.env.cr.commit()  # pylint: disable=E8102
```
**suggestion:** Maybe add a comment to explain why you can't use a different cursor (lock never released by Odoo)?

Koudja replied:
(doesn't seem useful anymore)

alexandregaldeano reviewed (approved):
Once the tests are fixed, LGTM! 🚀

Koudja committed:
[IMP] queue_job_cron_jobrunner: requeue lost started jobs

petrus-v reviewed (approved):
Nice one, thanks !

petrus-v commented:
This PR require https://github.com/OCA/queue/pull/754 to avoid opened file descriptor leak !

hbrunn commented:
@Koudja this is still in draft state, is this intentional?

petrus-v commented:
> @Koudja this is still in draft state, is this intentional?

In fact we are still supervising the system and keep improving it as long we met difficulties ! yesterday we've implement the stop_processing feature to avoid to be kill by the cpu time limit in case there is a lot of job in the queue, this is not yet deployed to our production (and I'm going to somme logging).

So I suppose keeping it in draft state as long we have not get some stabilities is probably better ! But review and feedback are welcome otherwise so not sure what suits most OCA conventions in such case ?

hbrunn commented:
draft is good I think, because of the label I nearly merged it

petrus-v commented on queue_job_cron_jobrunner/models/queue_job.py:
```diff
@@ -113,18 +162,47 @@ def _process(self, commit=False):
+                "queue_job_cron_jobrunner.stop_processing_threshold_seconds", "5"
+            )
+        )
+        if next_cron_job_runner_trigger.nextcall <= (
```
after few investigation, I've the feeling the nextcall is not yet uppdated so we are probably dealing with the current started date so we would have to compute it...

petrus-v replied:
because odoo changed it on a different cursor !...

sbidoul commented:
I'm wondering... now that the job scheduler (misnamed jobrunner) can run on odoo.sh (#668), queue_job_cron_jobrunner could maybe be simplified to only run jobs that are in state `enqueued`, and keep all the scheduling logic (channels, priorities, eta, etc) in the regular jobrunner ?

petrus-v commented:
> I'm wondering... now that the job scheduler (misnamed jobrunner) can run on odoo.sh (#668), queue_job_cron_jobrunner could maybe be simplified to only run jobs that are in state `enqueued`, and keep all the scheduling logic (channels, priorities, eta, etc) in the regular jobrunner ?

As of today, we are not using the mentioned patch and have configured our setup to avoid spawning the job scheduler/runner, as suggested in the module’s documentation.

To be honest, we haven’t spent much time evaluating this patch, so we might have missed some key information. However, we do have some concerns:

  *  A preference for using a cron worker to process queue jobs.
  *  Uncertainty about how Odoo.sh manages process lifecycles, especially regarding jobrunner. If it gets killed (due to inactivity, lack of HTTP requests, or other obscure platform limits), how does the system handle it? Would it be properly restarted by Odoo.sh?
  
If I understand correctly, your idea is to have the queue jobrunner focus on managing job states—ensuring that tasks are enqueued while respecting channel capacities—while letting the ir.cron execute them based on priority. This approach would remove the need for HTTP workers to process jobs.

That sounds like an interesting direction and avoid to duplicate the channel implementation logic. Would you be open to further designing this together or discussing the long-term vision for the module?

sbidoul commented:
> If I understand correctly, your idea is to have the queue jobrunner focus on managing job states—ensuring that tasks are enqueued while respecting channel capacities—while letting the ir.cron execute them based on priority. This approach would remove the need for HTTP workers to process jobs.

Yes, that's the idea.

> Uncertainty about how Odoo.sh manages process lifecycles, especially regarding jobrunner. If it gets killed (due to inactivity, lack of HTTP requests, or other obscure platform limits), how does the system handle it? Would it be properly restarted by Odoo.sh?

I'm not sure either. On the project where we use the HA jobrunner on odoo.sh, we don't observe issues with this, but yeah, "obscure platform limits" are the key words. But I tend to think this is not a blocker for that approach.

> That sounds like an interesting direction and avoid to duplicate the channel implementation logic. Would you be open to further designing this together or discussing the long-term vision for the module?

We don't need this so far :crossed_fingers:, but I'd be happy to discuss.

Koudja commented:
I updated PR with latest comments. It's been on our production for few weeks now without any issue (with thousands jobs per day). I updated this PR state to ready for merge based on this.

Koudja commented:
We've migrated on HTTP worker, so I am closing this PR.
